src/devices/acurite.c

```diff
@@ -713,39 +836,25 @@ static int acurite_txr_decode(r_device *decoder, bitbuffer_t *bitbuffer)
         // M = Message type
         message_type = bb[2] & 0x3f;
 
-        // tower sensor messages are 7 bytes.
-        // TODO: - see if there is a type in the message that
-        // can be used instead of length to determine type
+        // Multiple AcuRite sensors use the same basic message format,
+        // with a shared header and integrity scheme. These are differentiated
+        // by the message_type.
         if (browlen == ACURITE_TXR_BITLEN / 8) {
-            char const *channel_str = acurite_getChannel(bb[0]);
-            // Tower sensor ID is the last 14 bits of byte 0 and 1
-            // CCII IIII | IIII IIII
-            sensor_id = ((bb[0] & 0x3f) << 8) | bb[1];
-            //sensor_status = bb[2]; // TODO:, uses parity? & 0x07f
-            humidity = (bb[3] & 0x7f); // 1-99 %rH
-            // temperature encoding used by "tower" sensors 592txr
-            // 14 bits available after removing both parity bits.
-            // 11 bits needed for specified range -40 C to 70 C (-40 F - 158 F)
-            // range -100 C to 1538.4 C
-            int temp_raw = ((bb[4] & 0x7F) << 7) | (bb[5] & 0x7F);
-            tempc = temp_raw * 0.1 - 100;
-            // Battery status is the 7th bit 0x40. 1 = normal, 0 = low
-            battery_low = (bb[2] & 0x40) == 0;
+            int decoded = 0;
 
-            /* clang-format off */
-            data = data_make(
-                    "model",                "",             DATA_STRING, "Acurite-Tower",
-                    "id",                   "",             DATA_INT,    sensor_id,
-                    "channel",              NULL,           DATA_STRING, channel_str,
-                    "battery_ok",           "Battery",      DATA_INT,    !battery_low,
-                    "temperature_C",        "Temperature",  DATA_FORMAT, "%.1f C", DATA_DOUBLE, tempc,
-                    "humidity",             "Humidity",     DATA_FORMAT, "%u %%", DATA_INT,    humidity,
-                    "mic",                  "Integrity",    DATA_STRING, "CHECKSUM",
-                    NULL);
-            /* clang-format on */
+            if (message_type == ACURITE_MSGTYPE_LEAK_DETECTOR) {
+                decoded = acurite_leak_detector_decode(decoder, bb);
+            } else if (message_type == ACURITE_MSGTYPE_TOWER_SENSOR) {
+                decoded = acurite_tower_sensor_decode(decoder, bb);
+            }
 
-            decoder_output_data(decoder, data);
-            valid++;
+            // The decoder attempts for this size message will return a positive
+            // value if they successfully decoded a message.
+            if (decoded > 0) {
+                valid++;
+            } else {
+                continue;
+            }
```

Comment:
The changes look very nice. Honwever, one nit -- Please delete this continue. 

This is what I was trying to get rid of.  Unless I haven't had enough coffee yet, It shouldn't be necessary and creates some subtle complexity for maintenance (if you want to handle any 7 byte messages you need to do it here).

Reply:
Done

Reply:
I agree, this continue isn't necessary. If not removed, it could be placed after `valid++;` inside the if (if we decoded a message, move to the next one without evaluating the other decoders (515, 5n1)). But just nuking it is fine, probably premature optimization to try and avoid checking other decoders.